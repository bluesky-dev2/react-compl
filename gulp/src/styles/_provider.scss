.react-contexify-provider {
	display: inline-block;
}