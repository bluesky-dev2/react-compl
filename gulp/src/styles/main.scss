@charset "UTF-8";

// Layout ------------------------------------------------------------ //
@import "container";
@import 'provider';
@import 'menu';

//Themes ------------------------------------------------------------ //
@import "themes/dark";
@import "themes/light";

//Animations -------------------------------------------------------- //
@import "animations/popIn";
@import "animations/zoomIn";
@import "animations/fadeIn";
@import "animations/flipIn";
