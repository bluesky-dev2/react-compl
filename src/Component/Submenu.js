import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import cssClasses from './../cssClasses';

export default class Submenu extends PureComponent {
  static propTypes = {
    children: PropTypes.node.isRequired,
    title: PropTypes.string.isRequired,
    leftIcon: PropTypes.node,
    rightIcon: PropTypes.node,
    disabled: PropTypes.bool,
  };

  static defaultProps = {
    leftIcon: '',
    rightIcon: '',
    disabled: false,
    onClick: () => {},
    targetNode: {},
    data: null,
    refsFromProvider: []
  };

  render() {
    const className = cx(cssClasses.ITEM, {
      [`${cssClasses.ITEM_DISABLED}`]: this.props.disabled
    });

    return (
      <div className={className} role="presentation">
        <div className={cssClasses.ITEM_DATA}>
          {this.props.leftIcon}
          {this.props.title}
          {this.props.rightIcon}
          <span>â–¶</span>
        </div>
        <div className={cssClasses.SUBMENU}>{this.props.children}</div>
      </div>
    );
  }
}
