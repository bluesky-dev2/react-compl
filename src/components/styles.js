/*
* css classes mapping
* */
export default {
  menu: 'react-contexify',
  submenu: 'react-contexify react-contexify__submenu',
  submenuArrow: 'react-contexify__submenu-arrow',
  separator: 'react-contexify__separator',
  item: 'react-contexify__item',
  itemDisabled: 'react-contexify__item--disabled',
  itemContent: 'react-contexify__item__data',
  itemIcon: 'react-contexify__item__icon',
  theme: 'react-contexify__theme--',
  animationWillEnter: 'react-contexify__will-enter--',
  animationWillLeave: 'react-contexify__will-leave--'
};
