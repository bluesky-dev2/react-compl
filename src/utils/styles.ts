/*
 * css classes mapping
 * */
export const styles = {
  menu: 'react-contexify',
  submenu: 'react-contexify react-contexify__submenu',
  submenuArrow: 'react-contexify__submenu-arrow',
  separator: 'react-contexify__separator',
  item: 'react-contexify__item',
  itemDisabled: 'react-contexify__item--disabled',
  itemContent: 'react-contexify__item__content',
  itemIcon: 'react-contexify__item__icon',
  theme: 'react-contexify__theme--',
  animationWillEnter: 'react-contexify__will-enter--',
  animationWillLeave: 'react-contexify__will-leave--'
};

// I could use enum instead for both interfaces

export interface BuiltInTheme {
  light: string;
  dark: string;
}

export interface BuiltInAnimation {
  fade: string;
  flip: string;
  pop: string;
  zoom: string;
}

export const theme: BuiltInTheme = {
  light: 'light',
  dark: 'dark'
};

export const animation: BuiltInAnimation = {
  fade: 'fade',
  flip: 'flip',
  pop: 'pop',
  zoom: 'zoom'
};
