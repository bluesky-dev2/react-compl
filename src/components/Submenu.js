import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './styles';

export default class Submenu extends Component {
  static propTypes = {
    label: PropTypes.node.isRequired,
    children: PropTypes.node.isRequired,
    arrow: PropTypes.node,
    nativeEvent: PropTypes.object,
    dataFromProvider: PropTypes.any,
    refsFromProvider: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.object)
    ]),
    className: PropTypes.string,
    style: PropTypes.object,
    disabled: PropTypes.oneOfType([PropTypes.bool, PropTypes.func])
  };

  static defaultProps = {
    arrow: 'â–¶',
    nativeEvent: null,
    dataFromProvider: null,
    refsFromProvider: null,
    className: null,
    style: {},
    disabled: false
  };

  state = {
    style: {
      left: '100%',
      top:0,
      bottom: 'initial'
    }
  };

  setRef = ref => {
    this.menu = ref;
  };

  componentDidMount() {
    const { innerWidth, innerHeight } = window;
    const rect = this.menu.getBoundingClientRect();
    const style = {};

    if(rect.right < innerWidth) {
      style.left = '100%';
    } else {
      style.right = '100%';
    }

    if (rect.bottom > innerHeight) {
      style.bottom = 0;
      style.top = 'initial';
    } else {
      style.bottom = 'initial';
      style.top = 0;
    }

    this.setState({
      style: style
    });
  }

  getMenuItem() {
    const { nativeEvent, refsFromProvider, dataFromProvider } = this.props;

    return React.Children.map(this.props.children, item =>
      React.cloneElement(item, {
        nativeEvent,
        refsFromProvider,
        dataFromProvider
      })
    );
  }

  handleClick(e) {
    e.stopPropagation();
  }

  render() {
    const {
      arrow,
      disabled,
      className,
      style,
      label,
      nativeEvent,
      refsFromProvider,
      dataFromProvider,
    } = this.props;
    const cssClasses = cx(styles.item, className, {
      [`${styles.itemDisabled}`]:
        typeof disabled === 'function'
          ? disabled({
              event: nativeEvent,
              dataFromProvider,
              ref: refsFromProvider
            })
          : disabled
    });
    const submenuStyle = {
      ...style,
      ...this.state.style
    };

    return (
      <div
        className={cssClasses}
        role="presentation"
      >
        <div className={styles.itemContent} onClick={this.handleClick}>
          {label}
          <span className={styles.submenuArrow}>{arrow}</span>
        </div>
        <div
          className={styles.submenu}
          ref={this.setRef}
          style={submenuStyle}
        >
          {this.getMenuItem()}
        </div>
      </div>
    );
  }
}
