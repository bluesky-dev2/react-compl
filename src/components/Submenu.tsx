import React, { Component, ReactNode, SyntheticEvent } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { cloneItem } from './cloneItem';
import { styles } from '../utils/styles';
import {
  MenuItemEventHandler,
  TriggerEvent,
  StyleProps,
  InternalProps
} from '../types';

export interface SubMenuProps extends StyleProps, InternalProps {
  /**
   * Any valid node that can be rendered
   */
  label: ReactNode;

  /**
   * Any valid node that can be rendered
   */
  children: ReactNode;

  /**
   * Render a custom arrow
   */
  arrow: ReactNode;

  /**
   * Disable or not the `Submenu`. If a function is used, a boolean must be returned
   */
  disabled: boolean | ((args: MenuItemEventHandler) => boolean);
}

interface SubMenuState {
  left?: string | number;
  right?: string | number;
  top?: string | number;
  bottom?: string | number;
}

class Submenu extends Component<SubMenuProps, SubMenuState> {
  static propTypes = {
    label: PropTypes.node.isRequired,
    children: PropTypes.node.isRequired,
    nativeEvent: PropTypes.object,
    arrow: PropTypes.node,
    disabled: PropTypes.oneOfType([PropTypes.func, PropTypes.bool]),
    className: PropTypes.string,
    style: PropTypes.object
  };
  static defaultProps = {
    arrow: 'â–¶',
    disabled: false,
    nativeEvent: {} as TriggerEvent
  };

  state = {
    left: '100%',
    top: 0,
    bottom: 'initial'
  };

  menu!: HTMLElement;

  setRef = (ref: HTMLDivElement) => {
    this.menu = ref;
  };

  componentDidMount() {
    const { innerWidth, innerHeight } = window;
    const rect = this.menu.getBoundingClientRect();
    const style: SubMenuState = {};

    if (rect.right < innerWidth) {
      style.left = '100%';
      style.right = undefined;
    } else {
      style.right = '100%';
      style.left = undefined;
    }

    if (rect.bottom > innerHeight) {
      style.bottom = 0;
      style.top = 'initial';
    } else {
      style.bottom = 'initial';
      style.top = 0;
    }

    this.setState(style);
  }

  handleClick(e: SyntheticEvent) {
    e.stopPropagation();
  }

  render() {
    const {
      arrow,
      disabled,
      className,
      style,
      label,
      nativeEvent,
      children,
      propsFromTrigger
    } = this.props;

    const cssClasses = cx(styles.item, className, {
      [`${styles.itemDisabled}`]:
        typeof disabled === 'function'
          ? disabled({
              event: nativeEvent as TriggerEvent,
              props: { ...propsFromTrigger }
            })
          : disabled
    });

    const submenuStyle = {
      ...style,
      ...this.state
    };

    return (
      <div className={cssClasses} role="presentation">
        <div className={styles.itemContent} onClick={this.handleClick}>
          {label}
          <span className={styles.submenuArrow}>{arrow}</span>
        </div>
        <div className={styles.submenu} ref={this.setRef} style={submenuStyle}>
          {cloneItem(children, {
            propsFromTrigger,
            nativeEvent: nativeEvent as TriggerEvent
          })}
        </div>
      </div>
    );
  }
}

export { Submenu };
