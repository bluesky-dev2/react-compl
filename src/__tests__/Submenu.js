/* eslint-env jest */
import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Submenu from './../components/Submenu';
import styles from './../utils/styles';

const Children = () => <div>foo</div>;
describe('Submenu', () => {
  it('Should render without crash 💥', () => {
    const component = mount(
      <Submenu label="bar">
        <Children />
      </Submenu>
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it('Should not go offscreen', () => {
    global.innerWidth = 100;
    global.innerHeight = 100;
    Element.prototype.getBoundingClientRect = () => ({
      right: 200,
      bottom: 200
    });
    const component = mount(
      <Submenu label="bar">
        <Children />
      </Submenu>
    );

    expect(toJson(component)).toMatchSnapshot();
  });

  it('Can be disabled with true or a function', () => {
    let component = mount(
      <Submenu label="bar" disabled>
        <Children />
      </Submenu>
    );
    expect(component.html()).toMatch(styles.itemDisabled);

    component.unmount();
    component = mount(
      <Submenu label="bar" disabled={() => true}>
        <Children />
      </Submenu>
    );
    expect(component.html()).toMatch(styles.itemDisabled);
  });

  it('Should accept a custom arrow ➡️', () => {
    const component = mount(
      <Submenu label="bar" arrow="🦄">
        <Children />
      </Submenu>
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});
