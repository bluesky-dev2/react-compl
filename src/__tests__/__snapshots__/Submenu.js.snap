// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submenu Should not go offscreen 1`] = `
<Submenu
  className={null}
  dataFromProvider={null}
  disabled={false}
  label="bar"
  refsFromProvider={null}
  style={Object {}}
  targetNode={null}
>
  <div
    className="react-contexify__item"
    onClick={[Function]}
    role="presentation"
  >
    <div
      className="react-contexify__item__data"
    >
      bar
      <span>
        â–¶
      </span>
    </div>
    <div
      className="react-contexify react-contexify__submenu"
      style={
        Object {
          "bottom": 0,
          "left": "-100%",
          "top": "initial",
        }
      }
    >
      <Children
        dataFromProvider={null}
        key=".0"
        refsFromProvider={null}
        targetNode={null}
      >
        <div>
          foo
        </div>
      </Children>
    </div>
  </div>
</Submenu>
`;

exports[`Submenu Should render without crash ðŸ’¥ 1`] = `
<Submenu
  className={null}
  dataFromProvider={null}
  disabled={false}
  label="bar"
  refsFromProvider={null}
  style={Object {}}
  targetNode={null}
>
  <div
    className="react-contexify__item"
    onClick={[Function]}
    role="presentation"
  >
    <div
      className="react-contexify__item__data"
    >
      bar
      <span>
        â–¶
      </span>
    </div>
    <div
      className="react-contexify react-contexify__submenu"
      style={
        Object {
          "bottom": "initial",
          "left": "100%",
          "top": 0,
        }
      }
    >
      <Children
        dataFromProvider={null}
        key=".0"
        refsFromProvider={null}
        targetNode={null}
      >
        <div>
          foo
        </div>
      </Children>
    </div>
  </div>
</Submenu>
`;
